Cascade Layers (#2322)
withastro/starlight: docs/src/content/docs/guides/css-and-tailwind.mdx

Starlight uses [cascade layers](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers) internally so that any custom unlayered CSS will override the default styles.
If you are using cascade layers in your custom CSS, you can use the [`@layer`](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer) CSS at-rule to define the order of precedence for different layers including the ones used by Starlight:

```css /(my-(?:reset|layer))/
@layer starlight.base,
	my-reset,
	starlight.reset,
	starlight.components,
	starlight.content,
	starlight.user-components,
	starlight.asides,
	starlight.expressive-code,
	starlight.utils,
	my-layer;
```

## Tailwind CSS

Tailwind CSS support in Astro projects is provided by the [Tailwind Vite plugin](https://tailwindcss.com/docs/installation/using-vite).
Starlight provides complementary CSS to help configure Tailwind for compatibility with Starlight’s styles.

The Starlight Tailwind CSS applies the following configuration:

- Configures Tailwind’s `dark:` variants to work with Starlight’s dark mode.
- Uses Tailwind [theme colors and fonts](#styling-starlight-with-tailwind) in Starlight’s UI.

> - Restore essential parts of Preflight.

Reviewer: ```suggestion
- Restores essential parts of Preflight.
```

You can see all the CSS custom properties used by Starlight that you can set to customize your site in the [`props.css` file on GitHub](https://github.com/withastro/starlight/blob/main/packages/starlight/style/props.css).

### Cascade layers

> Starlight uses [cascade layers](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers) internally so that any custom unlayered CSS will override the default styles.

Reviewer 2: Small suggestion: This part can be slightly expanded to mention the main purpose of using layers and to include the `starlight` layer name.

```suggestion
Starlight uses [cascade layers](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers) internally for managing the order of styles. This results in a more predictable CSS order and allows for easier overrides. Any custom unlayered CSS will override the default styles from the `starlight` layer.
```

### Cascade layers

Starlight uses [cascade layers](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Cascade_layers) internally so that any custom unlayered CSS will override the default styles.
If you are using cascade layers in your custom CSS, you can use the [`@layer`](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer) CSS at-rule to define the order of precedence for different layers including the ones used by Starlight.

The following example shows how to use the `@layer` at-rule to define a custom layer named `my-layer` that will be applied after all Starlight layers and have precedence over them:

```css "my-layer"
@layer starlight, my-layer;
```

For more fine-grained control, you can also include your custom layers between Starlight’s layers.
The following example adds a custom layer named `my-reset` before Starlight’s `starlight.reset` layer and a custom layer named `my-layer` after all other Starlight layers:

```css /(my-(?:reset|layer))/
@layer starlight.base,
	starlight.reset,
	my-reset,
	starlight.core,
	starlight.content,
	starlight.components,
	starlight.utils,
	my-layer;

> ```

Reviewer 3: Thanks for catching that! Do people have opinions on how helpful that would be? I guess you could still add your reset like this?
```css
@layer starlight.base,
	starlight.reset,
	starlight.my-reset,
	starlight.core,
	starlight.content,
	starlight.components,
	starlight.utils,
	my-layer;
```
Or are top-level layers preferable? Something like
```css
@layer starlight-base,
	starlight-reset,
	my-reset,
	starlight-core,
	starlight-content,
	starlight-components,
	starlight-utils,
	my-layer;
```

> ```

Reviewer 2: I think I would do something like this:

```css
@layer my-reset, starlight, my-overrides;
```

Generally, it's best to have as few top-level layers as possible. Consumers should also not need to mess with Starlight's nested layers.

(I can share more feedback after I get the chance to play with it. Just trying to figure out how to do that)

> ```

Reviewer 3: > Just trying to figure out how to do that

Using https://gitpkg.vercel.app/ might be easiest?

1. Create a test project (or use Starlight site you already have set up
2. Update the Starlight dep:
	```sh
	npm install 'https://gitpkg.now.sh/HiDeoo/starlight/packages/starlight?hd-feat-cascade-layers'
	```
	
Alternatively you could just play around in the `docs` directory inside this monorepo using this branch.

> ```

Reviewer 4: > Thanks for catching that! Do people have opinions on how helpful that would be? I guess you could still add your reset like this? Although it would require you naming your layer with a `starlight.` prefix which wouldn’t work for everyone. 
> ```css
> @layer starlight.base,
> 	starlight.reset,
> 	starlight.my-reset,
> 	starlight.core,
> 	starlight.content,
> 	starlight.components,
> 	starlight.utils,
> 	my-layer;
> ```
> Or are top-level layers preferable? Something like
> ```css
> @layer starlight-base,
> 	starlight-reset,
> 	my-reset,
> 	starlight-core,
> 	starlight-content,
> 	starlight-components,
> 	starlight-utils,
> 	my-layer;
> ```

Because we already have some pre-existing layers in our component library I was trying to slot them into the right order but I was overthinking it somewhat and ended up solving it with a `.not-content` class.

I agree with @mayank99 and the top-level ones are fine and you can use your first approach if something specific needs doing (or just add your CSS to one of the existing `starlight` layers!)

> ```

Reviewer 4: > > Just trying to figure out how to do that
> 
> Using https://gitpkg.vercel.app/ might be easiest?
> 
> 1. Create a test project (or use a Starlight site you already have set up)
> 2. Update the Starlight dep:
> 	```sh
> 	npm install 'https://gitpkg.now.sh/HiDeoo/starlight/packages/starlight?hd-feat-cascade-layers'
> 	```
> 	
> Alternatively you could just play around in the `docs` directory inside this monorepo using this branch.

If you're using PNPM you can also do:

```json
"@astrojs/starlight": "github:markgaze/starlight#hd-feat-cascade-layers&path:/packages/starlight",
"@astrojs/starlight-tailwind": "github:markgaze/starlight#hd-feat-cascade-layers&path:/packages/tailwind",
```

in your `package.json` and it will work.

(I'm using my fork of the fork because I fixed the Tailwind package `files` as per my other review comment)

> ```

Reviewer 2: The `github:` approach sounds amazing! I'm pleasantly surprised to see there is no build step in Starlight packages

> ```

Reviewer 4: > The `github:` approach sounds amazing! I'm pleasantly surprised to see there is no build step in Starlight packages

I was too! If only I'd looked first and not wasted my time trying to `pnpm patch` them 🙈😂

The docs for that approach are [here](https://pnpm.io/cli/add#install-from-a-git-repository-combining-different-parameters), in case anyone needs them that sees this!

<Steps>

> 1.  Add Tailwind Vite plugin:

Reviewer: ```suggestion
1.  Add the Tailwind Vite plugin:
```

    </Tabs>

> 2.  Install the Starlight Tailwind CSS:

Reviewer 5: ```suggestion
2.  Install the Starlight Tailwind CSS plugin:
```

Author: I don't think `plugin` would be the proper word here as it could refer to what is now called "legacy plugins" in Tailwind CSS but what is being installed is indeed just CSS. Maybe "Install the Starlight Tailwind CSS declarations:" 🤔

Reviewer 5: How about `Add the Starlight Tailwind CSS rules:`?

The main confusion for me is "install". I would use it for plugins or something that has an Install process, but not otherwise. If this has an installation process then it's fine.

Author: In that case, there is still an installation process through `npm`, e.g. `npm install` which installs a package that only ships CSS. It's similar to [`normalize.css`](https://www.npmjs.com/package/normalize.css) which can also be installed through `npm` but only ships CSS.

    </Tabs>

3.  Create a CSS file for Tailwind’s base styles, for example at `src/tailwind.css`, including Starlight Tailwind CSS:

    ```css {4}
    /* src/tailwind.css */

>     @layer base, starlight.base, starlight.reset, starlight.core, starlight.content, starlight.components, starlight.utils, theme, components, utilities;

Reviewer 3: And again:

```suggestion
    @layer base, starlight, theme, components, utilities;
```

		file](#custom-css-styles) to apply this theme to your site.
	</Fragment>
	<Fragment slot="tailwind-docs">

> 		The example [Tailwind theme config](#styling-starlight-with-tailwind) below

Reviewer 3: I think “example” was intentional on my part here (although it’s a slightly clunky sentence) with the idea that the code shown was an _example_ of a Tailwind config with the generated theme palette etc. and that if you already had some config set, you probably wouldn’t just copy/paste the whole thing. (Unlike the plain CSS instructions above which just say “Add the following…”

Now that the Tailwind config is in CSS we _could_ potentially just show the variables for Tailwind too now?

````md

Add the following CSS variables to the `@theme` block in your Tailwind CSS file:

```css
/* Generated accent color palettes. */
--color-accent-200: ${dark['accent-high']};
--color-accent-600: ${light.accent};
/* etc. */
```
````
